gum/gumstalker.h

```diff
@@ -110,6 +110,118 @@ GUM_API void gum_stalker_unfollow (GumStalker * self, GumThreadId thread_id);
+ *
+ * This API has the following caveats:
+ *
+ * 1. This API is not marked GUM_API and is thus not available from JavaScript.
+ *    It is unlikely there will be scenarios where this is useful.
+ *
+ * 2. This API MUST be called from the thread which will be executed in the
+ *    child. Since blocks are cached in the GumExecCtx which is stored on a
+ *    per-thread basis and accessed through Thread Local Storage, it is not
+ *    possible to prefetch blocks into the cache of another thread.
+ *
+ * 3. This API should be called after gum_stalker_follow_me. It is likely that
+ *    the parent will wish to call gum_stalker_deactivate immediately after
+ *    following. Subsequently, gum_stalker_activate can be called within the
+ *    child after it is forked to start stalking the thread once more. The child
+ *    can then send blocks to be instrumented back to the parent via
+ *    inter-process communications. The parent can then call
+ *    gum_stalker_prefetch to instrument those blocks before forking the next
+ *    child. As a result of the fork, the child inherits a deactivated stalker
+ *    instance, thus both parent and child should release their stalker
+ *    instances upon completion if required.
+ *
+ * 4. Note that gum_stalker_activate takes a `target` pointer which is used to
+ *    allow stalker to be reactivated whilst executing in an excluded range and
+ *    guranatee that the thread is followed until the 'activation target'
+ *    address is reached. The simplest solution to this is to define an empty
+ *    function and then subsequently call it immediately after activation to
+ *    return stalker to its normal behaviour.
+ *
+ *    __attribute__((always_inline))
+ *    static inline void
+ *    activation_target (void)
+ *    {
+ *
+ *    }
+ *
+ * 5. Note that since both parent and child have an identical stalker instance,
+ *    they each have the exact same transformer. Since this transformer will
+ *    be used both to generate blocks to execute in the child and to prefetch
+ *    blocks in the parent, care should be taken to identify in which scenario
+ *    the transformer is operating. The parent will likely also transform and
+ *    execute a few blocks even if it is deactivated immediately afterwards.
+ *    Thus care should also be taken when any callouts are executed to determine
+ *    whether they are running the parent or child context.
+ *
+ * 6. For optimal performance, blocks should be prefetched `n` times, where `n`
+ *    is the trust threshold. Each time a block is executed, a cached copy
+ *    of the original bytes of the block is compared in order to detect
+ *    self-modifying code before a cached instrumented block is executed. The
+ *    trust threshold sets a limit of how many times the block should be
+ *    identical before this comparison step is skipped. Thus by reaching this
+ *    limit during prefetching, we avoid this comparison step in the child.
```

Comment:
Yeah. Let’s parametrise. The child should only perfetch if it has determined the block is not modified. Otherwise, multiple successive children may each partially increment the blocks recycle count to the trust threshold.